### ISSUE TITLE
Figure labels in latex not generated unless `fig.align` is set

### ISSUE THREAD

**Issue by mathematicalcoffee** [rstudio/rmarkdown#1400]
Summary - unsure if knitr or rmarkdown problem, but I do not get figure labels unless I set the `fig.align` or `out.width` option explicitly somewhere.

MWE

~~~
---
# MWE - first figure gets no \label but second does.
title: "Test"
output:
  pdf_document:
    fig_caption: true
    keep_tex: true
---

```{r setup, include=F}
knitr::opts_chunk$set(echo=F)
library(ggplot2)
```

A test document.

```{r asdf, fig.cap="A plot"}
ggplot(iris, aes(x=Sepal.Length, y=Petal.Width)) + geom_point()
knitr::opts_chunk$set(fig.align='center')
```

```{r fdas, fig.cap="The same plot"}
ggplot(iris, aes(x=Sepal.Length, y=Petal.Width)) + geom_point()
```
~~~

Relevant output of the tex file produced by `rmarkdown::render('test.rmd')`

~~~
\begin{document}
\maketitle

A test document.

\begin{figure}
\centering
\includegraphics{test_files/figure-latex/asdf-1.pdf}
\caption{A plot}
\end{figure}

\begin{figure}

{\centering \includegraphics{test_files/figure-latex/fdas-1} 

}

\caption{The same plot}\label{fig:fdas}
\end{figure}


\end{document}
~~~

Note how the first figure does not get a `\label{fig:asdf}` while the second figure does get a `\label{fig:fdas}`; if I remove the line

~~~
knitr::opts_chunk$set(fig.align='center')
~~~

then both figures do not get a label. I also tried changing `fig.align='center'` to `out.width='0.8\\textwidth'` and this produced a label for the second figure; but setting say `fig.width=8` did *not* produce a label for the second figure.

Must one of `fig.align` or `out.width` (presumably `out.height` and possibly other options) be set in order to get labelled figures? It is not obvious that this should be the case.

**Comment by yihui** [comment]
Yes.

Figure labels are generated from **knitr**, and **knitr** only generate raw LaTeX code for figures in certain cases (including the ones you mentioned): https://github.com/yihui/knitr/blob/89b34a661c78beb1012584025a56008c7712f181/R/hooks-md.R#L7-L13 In other cases, it generates Markdown `![]()` and Pandoc will convert Markdown to LaTeX. At the moment, Pandoc does not generate labels.

To force **knitr** to generate raw LaTeX for R Markdown documents, you may either set the chunk option `out.extra = ""`, or conditionally set `knitr::knit_hooks$set(plot = knitr::hook_plot_tex)` (conditional on the value of `knitr:::pandoc_to()`).

**Comment by yihui** [comment]
I have been hesitating for long on whether I should just use `hook_plot_tex` for R Markdown's LaTeX output. You are not the first one who has been confused.

Yeah, an SO question with an answer is a good idea. Thanks!

**Comment by jonathan-g** [comment]
FWIW, Pandoc will generate LaTeX labels from `![caption](image_file)` if you use [the Pandoc link attributes extension](https://pandoc.org/MANUAL.html#images) to add a label: 

```![this is my caption](my_image_file){#my-label}```

will generate

```
\begin{figure}
\hypertarget{my-label}{%
\centering
\includegraphics{my_image_file}
\caption{this is my caption}\label{my-label}
}
\end{figure}
```

so it would be possible for knitr/rmarkdown to produce LaTeX labels (and HTML anchors) for Pandoc markdown output without needing hooks to generate special raw LaTeX output

**Comment by cderv** [comment]
That is really interesting. It seems it has been included in pandoc since version 1.17
https://pandoc.org/releases.html#pandoc-1.17-20-mar-2016

Would you open another issue for a feature request so that we keep in mind that this could be implemented ? 

If would mean to pass `fig.cap` as an identifier for Markdown image conversion and let pandoc handles the captioning when converting to latex.
